src/many-modules/src/_9_account/features/multisig.rs

```diff
@@ -9,98 +9,168 @@ use many_macros::many_module;
+/// It is meant to be a note regarding a message, transaction, info or any
+/// type that requires meta information.
+#[derive(Default, Clone, Debug, PartialOrd, Eq, PartialEq)]
+pub struct Memo<const MAX_LENGTH: usize = MULTISIG_MEMO_DATA_MAX_SIZE> {
+    inner: Vec<MemoInner<MAX_LENGTH>>,
+}
+
+impl<const M: usize> Memo<M> {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn push_str(&mut self, str: String) -> Result<(), ManyError> {
+        self.inner.push(MemoInner::<M>::try_from(str)?);
+        Ok(())
+    }
+
+    pub fn push_byte_vec(&mut self, bytes: ByteVec) -> Result<(), ManyError> {
+        self.inner.push(MemoInner::<M>::try_from(bytes)?);
+        Ok(())
     }
 }
 
-impl Memo<String> {
-    // USED ONLY IN TESTS! NEVER REMOVE THIS GUARD! MEMOS SHOULDN'T BE PRODUCED FROM LARGE STRINGS
-    #[cfg(test)]
-    pub fn new(s: String) -> Self {
-        Memo(s)
+impl<const M: usize> From<MemoInner<M>> for Memo<M> {
+    fn from(inner: MemoInner<M>) -> Self {
+        Self { inner: vec![inner] }
     }
 }
 
-/// Data inside a transaction
-#[derive(Clone, Debug, PartialEq)]
-pub struct Data(ByteVec);
+impl<const M: usize> TryFrom<Either<String, ByteVec>> for Memo<M> {
+    type Error = ManyError;
 
-impl<C> Encode<C> for Data {
-    fn encode<W: encode::Write>(&self, e: &mut Encoder<W>, _: &mut C) -> Result<(), encode::Error<W::Error>> {
-        e.bytes(self.0.as_ref()).map(|_| ())
+    fn try_from(s: Either<String, ByteVec>) -> Result<Self, Self::Error> {
+        Ok(Self {
+            inner: vec![match s {
+                Either::Left(str) => MemoInner::String(str),
+                Either::Right(bstr) => MemoInner::ByteString(bstr),
+            }],
+        })
     }
 }
 
-impl<C> Decode<'_, C> for Data {
-    fn decode(d: &mut Decoder<'_>, _: &mut C) -> Result<Self, decode::Error> {
-        let b = d.bytes()?;
-        if b.len() > MULTISIG_MEMO_DATA_MAX_SIZE {
-            return Err(minicbor::decode::Error::message("Data size over limit"));
-        }
-        Ok(Data(b.to_vec().into()))
+impl<const M: usize> TryFrom<String> for Memo<M> {
+    type Error = ManyError;
+    fn try_from(s: String) -> Result<Self, Self::Error> {
+        Ok(Self::from(MemoInner::try_from(s)?))
     }
 }
 
-impl TryFrom<Vec<u8>> for Data {
-    type Error = String;
-    fn try_from(b: Vec<u8>) -> Result<Self, Self::Error> {
-        if b.len() > MULTISIG_MEMO_DATA_MAX_SIZE {
-            return Err(format!("Data size over limit {}", b.len()));
+impl<const M: usize> TryFrom<ByteVec> for Memo<M> {
+    type Error = ManyError;
+    fn try_from(b: ByteVec) -> Result<Self, Self::Error> {
+        if b.len() > M {
+            return Err(ManyError::unknown(format!(
+                "Data size ({}) over limit ({})",
+                b.len(),
+                M
+            )));
         }
-        Ok(Data(b.into()))
+        Ok(Self {
+            inner: vec![MemoInner::ByteString(b)],
+        })
```

Comment:
Isn't that already checked by the `MemoInner`, i.e., can't you simply do

```
Ok(Self::from(MemoInner::try_from(b)?))
```

Reply:
You're right. I did the `String` version but not this one. My bad.